datafusion/optimizer/src/simplify_expressions.rs

```diff
@@ -790,6 +791,37 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
                 right,
             } if !info.nullable(&left)? && left == right => lit(1),
 
+            //
+            // Rules for Modulo
+            //
+
+            // A % null --> null
+            BinaryExpr {
+                left: _,
+                op: Modulo,
+                right,
+            } if is_null(&right) => *right,
+            // null % A --> null
+            BinaryExpr {
+                left,
+                op: Modulo,
+                right: _,
+            } if is_null(&left) => *left,
+            // A % 1 --> 0
+            BinaryExpr {
+                left,
+                op: Modulo,
+                right,
+            } if !info.nullable(&left)? && is_one(&right) => lit(0),
+            // A % 0 --> DivideByZero Error
+            BinaryExpr {
+                left,
+                op: Modulo,
+                right,
+            } if !info.nullable(&left)? && is_zero(&right) => {
+                return Err(DataFusionError::ArrowError(ArrowError::DivideByZero))
```

Comment:
I suggest not erroring in this case. It might end up causing a an expression like

```sql
select 
  CASE 
    WHEN 0 = 0 THEN 5 
    WHEN 0 != 0 THEN 5 % 0
   END
   ...
```

In other words, it could generate an error that would not ever actually be evaluated at runtime

Reply:
I removed the check.

Reply:
Hi @alamb @askoa, thank you for bringing up the discussion about the error checking. Here is my opinion.
We should do as much error checking as possible at compile time, whatever the error will be evaluated or not. Because this is much cheaper than checking at runtime. We don't want to run a long query and panics at an error that could be checked at compile time, such as `A % literal(0)`.

In Postgresql, `A % literal(0)` is checked at compiled time: https://onecompiler.com/postgresql/3yhrjpgrr

Reply:
I agree with @HaoYang670.  It does make sense to stop users from writing wrong queries irrespective of its going to be executed at runtime or not. If `A % 0` is wrong then why not force the user to change the query at compile time rather than taking a chance to see if that expression gets executed at run time?

Reply:
I'll revert the changes tomorrow. Let me know if you have any issues.

Reply:
The example from postgres is compelling -- thank you. I am fine with making it error and I apologize for the distractions. 

The situation I was thinking about was when people wrote a query with a parameter like `?` which got expanded for each individually query.